code/mob/living/silicon/ai/holograms.dm

```diff
@@ -1,30 +1,111 @@
 //AI HOLOGRAMS
+
+#define HOLOGRAM_PICTURE 1
+#define HOLOGRAM_TEXT 2
+
+/datum/ai_hologram_data
+	var/image_expansion
+	var/holograms = 0
+	var/const/max_holograms = 8
+	var/text_expansion
+	var/text_holograms = 0
+	var/const/max_text_holograms = 3
+
+	proc/reserve(obj/hologram/H)
+		if(!istype(H))
+			return
+		switch(H.hologram_type)
+			if(HOLOGRAM_PICTURE)
+				src.holograms += H.hologram_value
+			if(HOLOGRAM_TEXT)
+				src.text_holograms += H.hologram_value
+
+	proc/free(obj/hologram/H)
+		if(!istype(H))
+			return
+		switch(H.hologram_type)
+			if(HOLOGRAM_PICTURE)
+				src.holograms = max(src.holograms - H.hologram_value, 0)
+			if(HOLOGRAM_TEXT)
+				src.text_holograms = max(src.text_holograms - H.hologram_value, 0)
+
+	proc/check(holotype_string, mob/dead/aieye/E)
+		. = TRUE
+		if(!istype(E))
+			. = FALSE
+		switch(holotype_string)
+			if("write")
+				if (src.text_holograms >= src.max_text_holograms)
+					boutput(E, "Not enough T-RAM to project more text holograms. Delete others to make room.")
+					. = FALSE
+			else
+				if (src.holograms >= src.max_holograms)
+					boutput(E, "Not enough V-RAM to project more holograms. Delete others to make room.")
+					. = FALSE
+
+	proc/expansion(string)
+		text_expansion = string
+
+
 /mob/living/silicon/ai
 	contextLayout = new /datum/contextLayout/experimentalcircle(36)
 
 	proc/create_hologram()
 		if (!deployed_to_eyecam)
 			boutput(src, "Deploy to an AI Eye first to create a hologram.")
 			return
-		if (length(holograms)>= max_holograms)
-			boutput(eyecam, "Not enough RAM to project more holograms. Delete others to make room.")
-			return
 
 		var/turf/T = get_turf(src.eyecam)
 		src.show_hologram_context(T)
 
-	proc/create_hologram_at_turf(var/turf/T, var/holo_type)
+	proc/create_hologram_at_turf(turf/T, holo_type)
 		if (!deployed_to_eyecam)
 			boutput(src, "Deploy to an AI Eye first to create a hologram.")
 			return
-		if (length(holograms)>= max_holograms)
-			boutput(eyecam, "Not enough RAM to project more holograms. Delete others to make room.")
-			return
+
 		if (!istype(T) || !istype(T.cameras) || T.cameras.len == 0)
 			boutput(eyecam, "No camera available to project a hologram from.")
 			return
 
-		new /obj/hologram(T, owner=src, holo_type=holo_type)
+		if(!src.holoHolder.check(holo_type, eyecam))
+			return
+
+		if(holo_type=="write")
+			var/list/holo_setences = list()
+			var/list/holo_actions = list()
+			var/list/holo_nouns = list()
+			holo_setences += strings("hologram.txt", "sentences")
```

Comment:
is this cached?

Reply:
Yep! As soon as `strings()` nom-noms a file it will be cached.